hkmc2/shared/src/main/scala/hkmc2/codegen/js/JSBuilder.scala

```diff
@@ -0,0 +1,76 @@
+package hkmc2
+package codegen
+package js
+
+import mlscript.utils.*, shorthands.*
+import utils.*
+import document.*
+
+import Scope.scope
+
+
+// TODO factor some logic for other codegen backends
+abstract class CodeBuilder:
+    
+  type Context
+  
+
+class JSBuilder extends CodeBuilder:
+  
+  val builtinOps: Set[Str] =
+    Set("+", "-", "*", "/", "%", "==", "!=", "<", "<=", ">", ">=", "&&", "||")
+
+  // TODO use this to avoid parens when we generate recomposed expressions later
+  enum Context:
+    case TopLevel
+    case SelectionPrefix
+    case Argument
+    case Operand(prec: Int)
+  
+  def result(r: Result)(using Raise, Scope): Document = r match
+    case Value.Ref(m: semantics.MemberSymbol[?]) =>
+      m.nme // TODO qualify if necessary? (maybe never needed)
+    case Value.Ref(l: semantics.BlockLocalSymbol) =>
+      summon[Scope].lookup_!(l)
+    case Value.Lit(lit) =>
+      lit.idStr
+    case Call(Value.Ref(l: semantics.MemberSymbol[?]), lhs :: rhs :: Nil) if builtinOps contains l.nme =>
+      doc"${result(lhs)} ${l.nme} ${result(rhs)}"
+    case Call(fun, args) =>
+      doc"${result(fun)}(${args.map(result).mkDocument(", ")})"
+    case Value.Lam(ps, bod) => scope.nest givenIn:
+      val vars = ps.map(p => scope.allocateName(p.sym)).mkDocument(", ")
+      doc"function($vars) { #{  # ${
```

Comment:
Is it better to use the arrow syntax for anonymous functions?

Reply:
Yeah probably, to avoid rebinding `this`. Currently classes are not supported so it doesn't make a ton of difference, though.